### PR TITLE
add file type inference

### PR DESCRIPTION
**Pull request by johanmazelanssi**

### CODE REVIEW

File: pcap-rewrite/src/lib.rs [hunk 1]
```diff
@@ -80,19 +85,60 @@ fn get_reader(input_filename: &str) -> io::Result<Box<dyn Read>> {
...
+        fn pcap_same_endianess_matcher(buf: &[u8]) -> bool {
+            buf.len() >= 4 && buf[0] == 0xa1 && buf[1] == 0xb2 && buf[2] == 0xc3 && buf[3] == 0xd4
         }
+        fn pcap_reverse_endianess_matcher(buf: &[u8]) -> bool {
+            buf.len() >= 4 && buf[0] == 0xd4 && buf[1] == 0xc3 && buf[2] == 0xb2 && buf[3] == 0xa1
+        }
+        let mut info = infer::Infer::new();
+        info.add("custom/lz4", "lz4", lz4_matcher);
+        info.add("custom/pcap", "pcap", pcap_same_endianess_matcher);
+        info.add("custom/pcap", "pcap", pcap_reverse_endianess_matcher);
```
Comment (chifflier) [inline pcap-rewrite/src/lib.rs:107]: The detection on pcap-ng files is missing here.
To identify a pcap-ng file, you should look for:
- `> 12` bytes
- first 4 bytes are equal to `SHB_MAGIC` (`0x0A0D_0D0A`)
- at offset 8, the 4 next bytes are `BOM_MAGIC` (`0x1A2B_3C4D`) or the endian-swapped version of `BOM_MAGIC`

File: pcap-rewrite/src/lib.rs [hunk 2]
```diff
@@ -80,19 +85,60 @@ fn get_reader(input_filename: &str) -> io::Result<Box<dyn Read>> {
...
+        file.read_exact(&mut buf)?;
+
+        let kind = info.get(&buf).unwrap();
+
+        file.seek(SeekFrom::Start(0))?;
+        let b: Result<Box<dyn Read>, Error> = match kind.mime_type() {
+            "application/gz" => {
+                if !input_filename.ends_with(".gz") {
+                    warn!("Inferred file type is gz but file extension is not gz")
+                }
```
Comment (chifflier) [inline pcap-rewrite/src/lib.rs:119]: Are these warnings useful? The new code would not rely on the extension at all
Unless you just want to inform the user that the extension is wrong - that could be a valid reason

**Review by chifflier — CHANGES_REQUESTED** [review]
Overall, PR is good, but support for pcap-ng is missing (see comment for details)
Thanks!

File: pcap-rewrite/src/lib.rs (hunk 2 above)
Comment (johanmazelanssi) [reply pcap-rewrite/src/lib.rs:119]: I indeed wanted to warn the user about the extension (cf personal experience :)).
Should I use info!() instead?

Comment (chifflier) [reply pcap-rewrite/src/lib.rs:155]: If this is intentional, I'm ok with it :)

**Comment by chifflier** [comment]
New version is good, thanks!

**Pull request merged**